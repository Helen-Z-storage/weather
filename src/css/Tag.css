
.tags button {
    background-color: #DCDCDC;
    padding: 0.5em;
    margin: 0.15em;
    border: hidden;
    border-radius: 4px;
  }

.tagError {
  color: red;
  }