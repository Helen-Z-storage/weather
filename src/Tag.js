import React from 'react';
import './Tag.css';
const tag_spliter = " ";

function Tag(props) {
    const {tags, handleFilter, setNewWeather} = props;

    // split string tag into a list of valid tags, and remove empty tag
    var tag_lst = tags.split(tag_spliter).filter(tag => tag !== "")
                  .map((tag, i) => <button key={i} onClick={e => handleFilter(e)}>{tag}</button>);
  
    return (
        <ul>
            <li key="tags" className='tags'>{tag_lst}</li>
            <li><input type="text" onKeyUp={event => setNewWeather(event)} placeholder="Add a tag" /></li>
        </ul>
    );
}

export default Tag;
