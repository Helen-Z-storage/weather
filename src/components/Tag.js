import React from 'react';
import '../css/Tag.css';

import { connect } from "react-redux";
import * as uiActions from "../redux/actions/uiActions";
import * as pageActions from "../redux/actions/pageActions";

const tag_spliter = " ";
const defaultCity = {
    id: -1,
    country: "",
    city: "",
    lat: 0.0,
    lon: 0.0,
    tags: tag_spliter,
    icon: "",
    avg_forecast: {
        timestamp: "",
        temp: 0.00,
        feels_like: 0.00,
        temp_min: 0.00,
        temp_max: 0.00,
        icon: ""
    },
    detail_forecast: new Array(8).fill({
        timestamp: "", 
        temp: 0.00,
        feels_like: 0.00,
        temp_min: 0.00,
        temp_max: 0.00,
        icon: "",
    })
};

function Tag(props) {
    const {cityID} = props;
    let capitals = props.weather.getIn("weather.captials".split("."));
    const city = capitals[cityID] || defaultCity;

    const handleFilter = ({ target }) => {
      props.dispatch(uiActions.setTagFilter(target.innerHTML.toUpperCase()));
    }

    const updateWeatherTag = ({target, nativeEvent}) => {
      let clear = false;
      // handle case of tag_spliter exist in input value
      if (target.value.indexOf(tag_spliter) > -1) {
        alert(`No "${tag_spliter}" should include in Tag`);
        clear = true;
      }

      // handle case of invalid empty tag
      else if (nativeEvent.keyCode === 13 && target.value === "") {
          alert(`No empty tag should been add`);
          clear = true;
      }

      // saving tag iff press "enter"
      else if (nativeEvent.keyCode === 13) {
          var tag = target.value + tag_spliter;
          // Add new tag if tag is not store in current city
          if (city.tags.indexOf(tag_spliter + tag) === -1) {
            city.tags += (tag);
          } else {
              // handle case of duplicate tag
              alert(`No duplicate tag should been add`);
          }
          clear = true;
      }

      // clear text input field
      if (clear) {
          target.value = "";
      };

      props.dispatch(pageActions.pageSetWeather(capitals));
    };

    // split string tag into a list of valid tags, and remove empty tag
    var tag_lst = city.tags.split(tag_spliter).filter(tag => tag !== "")
                  .map((tag, i) => <button key={i} onClick={event => handleFilter(event)}>{tag}</button>);
  
    return (
        <ul>
            <li key="tags" className='tags'>{tag_lst}</li>
            <li><input type="text" onKeyUp={event => updateWeatherTag(event)} placeholder="Add a tag" /></li>
        </ul>
    );
}

export default connect (
    (state) => {
      return {
        ui: state.ui,
        weather: state.weather
      }
    })(Tag);
