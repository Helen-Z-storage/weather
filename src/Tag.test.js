import React from 'react'
import { create } from "react-test-renderer";
import { render, screen } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect' // expect function
import userEvent from '@testing-library/user-event'; 

import Tag from './Tag';

// required prop data of create component
const tags = " ad ver tise Ment 339%? ";
const tag_spliter = " ";
const enter = "{enter}";
const one_char_tag = "a" + enter;
const one_char_length = one_char_tag.length - enter.length + 1;
const multi_char_tag = "A@1+" + enter;
const multi_char_length = multi_char_tag.length - enter.length + 1;
const handleFilter = jest.fn();
const setNewStudent = jest.fn();
const setNewStudent1 = jest.fn();
const setNewStudent2 = jest.fn();

describe('testing Tag component', () => {

    test("Matches the snapshot", () => {
        // create component
        const tagData = create(
            <Tag tags={tags} handleFilter={handleFilter}
                setNewStudent={setNewStudent}/>
        );

        // expecting output
        expect(tagData.toJSON()).toMatchSnapshot();
    });
    
    test("shows clicked button working", () => {
        // render component
        render(
            <Tag tags={tags} handleFilter={handleFilter}
                 setNewStudent={setNewStudent}/>
        );

        // expecting output
        for (var tag of tags.split(tag_spliter).filter(tag => tag !== "")) {
            const button = screen.getByText(tag);
            userEvent.click(button); 
        }
        expect(handleFilter).toBeCalledTimes(5);
        expect(handleFilter.mock.calls.length).toEqual(5);
        expect(handleFilter.mock.calls[0].length).toEqual(1);
    });

    test("shows tag textfield working: one char tag", () => {
        // render component
        render(
            <Tag tags={tags} handleFilter={handleFilter}
                 setNewStudent={setNewStudent1}/>
        );
        
        // expecting output
        const textfield = screen.getByPlaceholderText("Add a tag");

        // this call setNewStudent1() one time with event contain "a" 
        userEvent.type(textfield, one_char_tag);
        expect(setNewStudent1).toBeCalledTimes(one_char_length);
        expect(setNewStudent1.mock.calls.length).toEqual(one_char_length);
        expect(setNewStudent1.mock.calls[0][0].target.value).toEqual("a");
    });

    test("shows tag textfield working: multi char tag", () => {
        // render component
        render(
            <Tag tags={tags} handleFilter={handleFilter}
                 setNewStudent={setNewStudent2}/>
        );
        
        // expecting output
        const textfield = screen.getByPlaceholderText("Add a tag");

        // call function 4 more times with event contain "A", "A@", "A@1", "A@1+"
        userEvent.type(textfield, multi_char_tag);
        const sum = multi_char_length;
        expect(setNewStudent2).toBeCalledTimes(sum);
        expect(setNewStudent2.mock.calls.length).toEqual(sum);
        expect(setNewStudent2.mock.calls[sum - 1][0].target.value).toEqual("A@1+");

    });
});