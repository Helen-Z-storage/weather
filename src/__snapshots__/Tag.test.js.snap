// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Tag component Matches the snapshot 1`] = `
<ul>
  <li
    className="tags"
  >
    <button
      onClick={[Function]}
    >
      ad
    </button>
    <button
      onClick={[Function]}
    >
      ver
    </button>
    <button
      onClick={[Function]}
    >
      tise
    </button>
    <button
      onClick={[Function]}
    >
      Ment
    </button>
    <button
      onClick={[Function]}
    >
      339%?
    </button>
  </li>
  <li>
    <input
      onKeyUp={[Function]}
      placeholder="Add a tag"
      type="text"
    />
  </li>
</ul>
`;
